include/mapbox/geojson_value_impl.hpp

```diff
@@ -0,0 +1,295 @@
+        }
+
+        feature_collection collection;
+        collection.reserve(featureArray->size());
+        for (const auto &featureValue : *featureArray) {
+            collection.push_back(convert<feature>(featureValue));
+        }
+        return geojson{ collection };
+    }
+
+    if (typeString == "Feature") {
+        return geojson{ convert<feature>(val) };
+    }
+
+    return geojson{ convert<geometry>(val) };
+}
+
+geojson convert(const value &val) {
+    return val.match(
+        [](const null_value_t &) -> geojson { return geometry{}; },
+        [](const std::string &jsonString) {
+            return jsonString == "null" ? geometry{} : parse(jsonString);
+        },
+        [](const value::object_type &jsonObject) {
+            return convert<geojson>(static_cast<const mapbox::geojson::value &>(jsonObject));
+        },
+        [](const auto &) -> geojson { throw error("Invalid GeoJSON value was provided."); });
+}
+
+value convert(const point &p) {
+    return value::array_type{ p.x, p.y };
+}
+
+template <typename Cont>
+value convert(const Cont &points) {
+    value::array_type result;
+    result.reserve(points.size());
+    for (const auto &p : points) {
+        result.push_back(convert(p));
+    }
+    return result;
+}
+
+value convert(const geometry &geom) {
+    return geom.match(
+        [](const empty &) { return value{}; },
+        [](const point &p) {
+            return value::object_type{ { "type", "Point" }, { "coordinates", convert(p) } };
+        },
+        [](const multi_point &mp) {
+            return value::object_type{ { "type", "MultiPoint" }, { "coordinates", convert(mp) } };
+        },
+        [](const line_string &ls) {
+            return value::object_type{ { "type", "LineString" }, { "coordinates", convert(ls) } };
+        },
+        [](const multi_line_string &mls) {
+            return value::object_type{ { "type", "MultiLineString" },
+                                       { "coordinates", convert(mls) } };
+        },
+        [](const polygon &pol) {
+            return value::object_type{ { "type", "Polygon" }, { "coordinates", convert(pol) } };
+        },
+        [](const multi_polygon &mpol) {
+            return value::object_type{ { "type", "MultiPolygon" },
+                                       { "coordinates", convert(mpol) } };
+        },
+        [](const geometry_collection &gc) {
+            value::array_type geometries;
+            geometries.reserve(gc.size());
+            for (const auto &gcGeom : gc) {
+                geometries.push_back(convert(gcGeom));
+            }
+            return value::object_type{ { "type", "GeometryCollection" },
+                                       { "geometries", std::move(geometries) } };
+        });
+}
+
+value convert(const feature &f) {
+    value::object_type result{ { "type", "Feature" }, { "geometry", convert(f.geometry) } };
+
+    if (!f.properties.empty()) {
```

Comment:
ditto, you need to add an empty `properties` at least

Reply:
done